### PR TITLE
Block API: Allow "array of attributes to be compared" for isActive property in block variations

### LINKED ISSUE #30739: Blocks: Make it more flexible to match the currently applied block variation
## What problem does this address?

It is possible to define multiple variations of the same block. Users can see in the UI a title, description, or icon if the currently applied block variation implements `isActive` callback and it returns `true`. 

> -   `isActive` (optional, type `Function`) - A function that accepts a block's attributes and the variation's attributes and determines if a variation is active. This function doesn't try to find a match dynamically based on all block's attributes, as in many cases some attributes are irrelevant. An example would be for `embed` block where we only care about `providerNameSlug` attribute's value.
>
> It's worth mentioning that setting the `isActive` property can be useful for cases you want to use information from the block variation, after a block's creation. For example, this API is used in `useBlockDisplayInformation` hook to fetch and display proper information on places like the `BlockCard` or `Breadcrumbs` components.

See more in: https://github.com/WordPress/gutenberg/blob/trunk/docs/reference-guides/block-api/block-variations.md

However, it looks like the majority of current usage loops over the list of variations and dynamically injects the `isActive` callback. Some examples:

https://github.com/WordPress/gutenberg/blob/bbd333386b022778ccb57c83a7ead68a460a94ab/packages/block-library/src/embed/variations.js#L342-L352

https://github.com/WordPress/gutenberg/blob/bbd333386b022778ccb57c83a7ead68a460a94ab/packages/block-library/src/social-link/variations.js#L307-L316

With #29095, it is also possible to register block variations on the server. It is impossible to define `isActive` in PHP as long as it needs to be represented as a JavaScript callback. It ends up being dynamically injected in JavaScript with the WordPress hook that creates another level of complexity.

https://github.com/WordPress/gutenberg/blob/b620c92c1bb0bb2097734ddaf3983de08b1318b4/packages/block-library/src/navigation-link/hooks.js#L47-L66

### Option 1: `isActive` allows arrays

This is a simple idea where `isActive` could be represented as an array in the case where block variations are registered on the server, e.g.:

```php
array(
	'name'     => 'category', 
	'isActive' => array( 'type' ),
	// ... other settings
)`
```

`type` is the value of a block attribute in the editor and the `type` set as an attribute for the variation. You can think of it as the following condition:

```js
isActive = ( blockAttributes, variationAttributes ) => blockAttributes.type === variationAttributes.type; 
```

### Options 2: top-level function on the client

Another option is that we allow defining the top-level API method that returns the name of the currently applied variation. In fact, this is what @ntsekouras originally proposed in #27469 when working on the API, but I convinced him to start with a callback per block variation so we keep the list of top-level settings small.

```js
registerBlockType( 'core/navigation-link', {
	getCurrentlyAplliedVariation: ( blockAttributes, blockVariations ) => {
		const { name } = find( blockVariations,  ( { attributes } ) => blockAttributes.type === attributes.type );
		return name;
 	},
	// ... other settings
} );
```

We could ensure backward compatibility by giving the higher priority the `isActive` matches defined on individual block variations.

### PR DESCRIPTION
**Pull request by david-szabo97**
Fixes https://github.com/WordPress/gutenberg/issues/30739
## Description
Extends the variations API by adding one more possible way to define how active variations are recognized.

The only way currently in `trunk` is to use an `isActive` `function`. In this PR we introduce `array of strings` for `isActive` property.

Most of our variations use the following pattern:

https://github.com/WordPress/gutenberg/blob/4b14897bc805ebe68093fca486b177eaa7e4513b/packages/block-library/src/embed/variations.js#L347-L352

As you can see, we are just creating a function for each variation and check for a specific attribute whether it matches or not. Basically, this is what we do for (almost) all of our variations so far.

To reduce boilerplate code, we can now use an array of strings to do the same check behind the scenes:

```js
variations.forEach( ( variation ) => {
	if ( variation.isActive ) return;
	variation.isActive = [ 'providerNameSlug' ];
} );
```

Every element in the array represents an attribute to be compared. If each attribute matches then it is considered active. For example:

```js
isActive: [ 'theme', 'level' ]
```

translates to

```js
( blockAttributes, variationAttributes ) =>
		blockAttributes.theme === variationAttributes.theme &&
		blockAttributes.level === variationAttributes.level ;
```

## How has this been tested?
1. Tests should pass
2. Smoke test, make sure all blocks using variations are still working as before

3.
Replace
https://github.com/WordPress/gutenberg/blob/4b14897bc805ebe68093fca486b177eaa7e4513b/packages/block-library/src/social-link/variations.js#L314
with
```js
	variation.isActive = [ 'service' ];
```
Make sure Social Link block still works the same way as before.

## Types of changes
New feature (non-breaking change which adds functionality)

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [ ] My code follows the accessibility standards. 
- [ ] I've tested my changes with keyboard and screen readers. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal).

### CODE REVIEW

File: docs/reference-guides/block-api/block-variations.md
Comment (ntsekouras) [inline docs/reference-guides/block-api/block-variations.md]: I think we can improve this with a starting small statement that accepts two types (function, array) and keep the rest of the text.

Also in added `string[]` in the description we should just say an array of attribute names as strings and state the way they will be used/compared (strict equality).

File: packages/blocks/src/store/selectors.js
Comment (ntsekouras) [inline packages/blocks/src/store/selectors.js]: If this selector ends up not exposed as stable, we should rename to `experimental...` to avoid all the lint disabled comments.. 

But I think we could make this stable, no? @gziolo

File: packages/blocks/src/store/selectors.js
Comment (ntsekouras) [inline packages/blocks/src/store/selectors.js:99]: I think we can expand a bit on the description.

File: packages/blocks/src/store/selectors.js
Comment (ntsekouras) [inline packages/blocks/src/store/selectors.js]: This is not needed.

File: packages/blocks/src/store/selectors.js
Comment (ntsekouras) [inline packages/blocks/src/store/selectors.js:121]: We need to validate in some place the provided as strings `attributes`, because in the case of having a non existing attribute the equality results to `undefined === undefined`, which is `true` and will show the first declared variation as `active`.

**Review by ntsekouras — COMMENTED** [review]
Thanks for working on this @david-szabo97! 💯 
 
In general this looks good. I haven't checked the tests you wrote for now, but I will after addressing the main thing about attributes validation.

File: packages/blocks/src/store/selectors.js
Comment (gziolo) [reply packages/blocks/src/store/selectors.js]: It all depends if we think that active block variation is clear enough for developers using this API. 😄

I assume that this API is here to stay so if we want to keep it experimental it should be:

`__experimentalGetActiveBlockVariation`

This way it will be omitted in documentation and ESLint won't complain.

File: packages/blocks/src/api/registration.js
Comment (gziolo) [inline packages/blocks/src/api/registration.js]: I guess we don't need this wrapper and we can use the selector from the store directly since it's used only in one place and it doesn't look like something that plugin developers would want to use very often.

File: packages/blocks/src/store/test/selectors.js
Comment (gziolo) [inline packages/blocks/src/store/test/selectors.js:388]: Jest has API for multiple variations of the same test:

https://jestjs.io/docs/api#testeachtablename-fn-timeout

File: packages/blocks/src/store/selectors.js
Comment (gziolo) [inline packages/blocks/src/store/selectors.js]: One thing to consider is whether we could combine `blockName` and `attributes` into a single param. 

```js
export function unstable__getActiveBlockVariation( state, { name, attributes}, scope );
```

With that instead of doing:

```js
const { attributes, name: blockName } = block;
const match = getActiveBlockVariation( blockName, attributes );
```

we could do:
```js
const match = getActiveBlockVariation( block );
```

**Review by gziolo — COMMENTED** [review]
Great work @david-szabo97, this is looking really good and nicely improves the API. We need to clarify the public API but otherwise it's exactly what we discussed 💯

File: packages/blocks/src/store/selectors.js
Comment (gwwar) [inline packages/blocks/src/store/selectors.js]: Optional: Would it be worth validating this earlier during block registration? It might be nice for folks to know that something isn't quite right when developing blocks.

File: packages/blocks/src/store/selectors.js
Comment (gwwar) [inline packages/blocks/src/store/selectors.js]: I think we'll still want to return false if either of these values is undefined. Blocks and variations may have optional attributes.

**Review by gwwar — COMMENTED** [review]
This looks pretty close @david-szabo97! I think this should be good to go once PR checks are ✅ and folks' feedback is addressed.

File: packages/blocks/src/store/selectors.js
Comment (david-szabo97) [reply packages/blocks/src/store/selectors.js]: I can't come up with a case for this 🤔 It seems to be working as expected:

1. If both of them are `undefined`, then that's correct and should match.
2. If `variation.attributes[ attribute ] = undefined` and `attributes[ attribute ] != undefined` then it won't match
3. If `variation.attributes[ attribute ] != undefined` and `attributes[ attribute ] = undefined` then it won't match

File: packages/blocks/src/store/selectors.js
Comment (david-szabo97) [reply packages/blocks/src/store/selectors.js]: Do we have any other instances where we do the same - block name + attributes under one object? So far I've only seen them separately like one param `block`/`blockName` and one param `attributes`.

File: packages/blocks/src/store/selectors.js
Comment (ntsekouras) [reply packages/blocks/src/store/selectors.js]: Actually that's what I had in mind too in my first comment. This way we will avoid calling this check so many times.

File: packages/blocks/src/store/selectors.js
Comment (ntsekouras) [reply packages/blocks/src/store/selectors.js]: @gwwar can you expand on the use case you have in mind with an example?

File: packages/blocks/src/store/selectors.js
Comment (gziolo) [reply packages/blocks/src/store/selectors.js]: Let's leave it as is then 👍🏻

File: packages/blocks/src/store/selectors.js
Comment (gwwar) [reply packages/blocks/src/store/selectors.js]: Sorry, I missed 👀 that we were already iterating over variations instead of all possible block attributes, so only the left side may be undefined, eg attributes[ attribute ]. Code here is fine 👍 

btw I personally find that code reuse in tests isn't that useful. What's nicer for me is being able to read a single test case without needing to skip to different definitions. (Fixtures/snapshots can also help when things go crazy). Curious if folks had thoughts on that.

Here's a free test case, where I verified this: 

```js
it( 'should handle optional attributes', () => {
				const testLinkName = 'test/link-block';
				const state = {
					blockTypes: {
						[ testLinkName ]: {
							name: testLinkName,
							attributes: {
								url: { type: 'string' },
								type: { type: 'string' },
								id: { type: 'number' },
							},
						},
					},
					blockVariations: {
						[ testLinkName ]: [
							{
								attributes: { url: 'https://wordpress.org' },
								isActive: [ 'url' ],
							},
							{
								attributes: { type: 'foo' },
								isActive: [ 'type' ],
							},
							{
								attributes: {
									type: 'bar',
									url: 'https://example.com',
								},
								isActive: [ 'type', 'url' ],
							},
						],
					},
				};

				expect(
					getActiveBlockVariation( state, testLinkName, {
						type: 'foo',
					} )
				).toEqual( {
					attributes: { type: 'foo' },
					isActive: [ 'type' ],
				} );

				expect(
					getActiveBlockVariation( state, testLinkName, {
						url: 'https://wordpress.org',
					} )
				).toEqual( {
					attributes: { url: 'https://wordpress.org' },
					isActive: [ 'url' ],
				} );

				expect(
					getActiveBlockVariation( state, testLinkName, {
						type: 'bar',
						url: 'https://example.com',
					} )
				).toEqual( {
					attributes: {
						type: 'bar',
						url: 'https://example.com',
					},
					isActive: [ 'type', 'url' ],
				} );

				expect(
					getActiveBlockVariation( state, testLinkName, {
						id: 1234,
					} )
				).toEqual( undefined );
			} );
```

Comment (david-szabo97) [reply packages/blocks/src/store/selectors.js]: Yeah, I had that in mind too. 😄 That's why the last tests are totally separate. Definitely needs some cleaning. A little bit of reuse doesn't hurt though 😄

Comment (gwwar) [reply packages/blocks/src/store/selectors.js]: > A little bit of reuse doesn't hurt though

Certainly, I wanted to leave a note since I was having a bit of trouble reading what the state shape was without inspecting tests.

**Review by gwwar — APPROVED** [review]
This looks solid to me @david-szabo97 ✨ Thanks for taking this one on!

I found another area where we should update before landing, which is the type definition in 

https://github.com/WordPress/gutenberg/blob/3e14c5e812d8077f28e080522f20763bbfb1dccc/packages/blocks/src/api/registration.js#L104-L111

🤔 Optional/followup: I wonder if there's a way to get this typedef to help generate the block-variations.md. We do have support for autogenerating other examples with `<!-- START TOKEN(Autogenerated API docs) -->` and `npm run docs:build`

@gziolo @ntsekouras were there any other blocking changes needed here?

**Comment by gziolo** [comment]
@gwwar, I will do a cross-check shortly. There is one follow-up that I can think of, it's changing API schema for block types in WordPress core to account for isActive as an array. In addition to that, we should be able to know to define `isActive` for Navigation Link block variations on the server 🎉

File: packages/blocks/src/store/selectors.js
Comment (gziolo) [inline packages/blocks/src/store/selectors.js]: I think `void` is incorrect here. It's ~probably~ according to the unit test `undefined` when the matching active variation is not found.

File: packages/blocks/src/store/selectors.js
Comment (gziolo) [reply packages/blocks/src/store/selectors.js:121]: Should we bail out early when there is an empty array passed as `isActive`?

**Review by gziolo — APPROVED** [review]

File: packages/blocks/src/store/selectors.js
Comment (ntsekouras) [inline packages/blocks/src/store/selectors.js]: We need to check if `.filter( ( attribute ) => attributeKeys.includes( attribute ) )` returns empty `array`.
If it does, in case of non existing attributes we must return `false` and bail early.

Now it will return `true` as this is how `every` works, resulting in returning the first declared block variation (https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/every).

>Caution: Calling this method on an empty array will return true for any condition!

**Review by ntsekouras — CHANGES_REQUESTED** [review]
Thanks @david-szabo97 ! A last change that is needed and let's land.

File: packages/blocks/src/store/selectors.js
Comment (david-szabo97) [reply packages/blocks/src/store/selectors.js]: I'll mark this as resolved. Renamed it to be stable.

File: packages/blocks/src/store/selectors.js
Comment (david-szabo97) [reply packages/blocks/src/store/selectors.js]: Fixed in https://github.com/WordPress/gutenberg/pull/30913/commits/66af876aef4138faad1fd07762dd5a26a6495b88

File: packages/blocks/src/store/selectors.js
Comment (david-szabo97) [reply packages/blocks/src/store/selectors.js]: Thanks! This will be obsolete once we validate variations at the time of block type registration.

File: packages/blocks/src/api/registration.js
Comment (david-szabo97) [inline packages/blocks/src/api/registration.js]: I'm not really sure where to put this. 😄 I'm also wondering why don't we `dispatch( blocksStore ).addBlockVariation( settings.variations )` instead of handling variations here:

https://github.com/WordPress/gutenberg/blob/6ad2a433769a4514fc52083e97aa47a0bc9edf07/packages/blocks/src/store/reducer.js#L139

For this reason we need to do the variations validation in two functions.

File: packages/blocks/src/store/selectors.js
Comment (david-szabo97) [reply packages/blocks/src/store/selectors.js]: Good point! First iteration is here: https://github.com/WordPress/gutenberg/pull/30913/commits/637b62bbd79f78daeece9f1ca6c7f89a552a0cff

File: packages/blocks/src/api/registration.js
Comment (gziolo) [reply packages/blocks/src/api/registration.js]: Can we extract it to another PR? It isn't a blocker really and it requires some more thinking because you can register block variations also through the store directly. I don't remember exactly but most likely they are stored in two places in the store. It's similar to how block styles can be registered, so plugin authors could register styles or variations before the block gets registered so plugins wouldn't have to declare dependencies on each other.

File: docs/reference-guides/block-api/block-variations.md
Comment (david-szabo97) [reply docs/reference-guides/block-api/block-variations.md]: Addressed in the latest commit.

File: packages/blocks/src/store/selectors.js
Comment (david-szabo97) [reply packages/blocks/src/store/selectors.js:99]: Addressed in the latest commit.

File: packages/blocks/src/api/registration.js
Comment (david-szabo97) [reply packages/blocks/src/api/registration.js]: That makes sense. I didn't know it meant to be that way - registering variation even without an existing block type. This comes with its own challenges 😄 

I revert the commit.

Comment (gziolo) [reply packages/blocks/src/api/registration.js]: I double-checked, it's stored in one reducer but you can add variations with `ADD_BLOCK_TYPES` or `ADD_BLOCK_VARIATIONS` 😄

**Review by ntsekouras — APPROVED** [review]
Let's 🚢 💯

**Pull request merged**
